### ISSUE TITLE
Use GaudiExec in all Ganga lessons

### ISSUE THREAD

**Issue by cmarinbe** [lhcb/first-analysis-steps#203]
This is the current recommendation so we should update the lessons to use this. We have been discussing, though, about the necessity to create a `lb-dev` folder even if you want to run from a released version. In principle, `prepareGaudiExec` should create it for you. We need to figure out how this exactly works.

**Comment by cmarinbe** [comment]
I've tried to do: 

```
j = Job()
myApp = prepareGaudiExec('DaVinci', 'v42r0')
myApp.options = ['code/davinci-grid/ntuple_options_grid.py']
j.application = myApp
```

the second line returns:
```
INFO     Set up App Env at: /afs/cern.ch/user/c/cmarinbe/cmtuser/DaVinciDev_v42r0
```
which seems in line of the folder being created automatically. 

However, when trying to add the input data as we were doing so far:
```
j.inputdata = j.application.readInputData('data/MC_2012_27163003_Beam4000GeV2012MagDownNu2.5Pythia8_Sim08e_Digi13_Trig0x409f0045_Reco14a_Stripping20NoPrescalingFlagged_ALLSTREAMS.DST.py')
```
an error raises:
```
INFO     Started parsing input Data file
ERROR    PythonOptsCmakeParserError:
GangaException:  ("The given directory: '/afs/cern.ch/user/c/cmarinbe/cmtuser/DaVinciDev_v42r0' doesn't exist!",)
ERROR    Error: ApplicationConfigurationError: Unable to parse the job options. Please check options files and extraopts. 
```
I've checked the folder is really not there. I guess it is only created when the job is prepared. I'm not sure why it is needed in order to parse the input data (project dependent?) but we need to figure out what's the best way off doing this now.

**Comment by cmarinbe** [comment]
Ok, my bad, the problem was the project version I was setting. Using: 
```
j = Job()
myApp = prepareGaudiExec('DaVinci','v40r2')
myApp.options = ['code/davinci-grid/ntuple_options_grid.py']
j.application = myApp
j.inputdata = j.application.readInputData('data/MC_2012_27163003_Beam4000GeV2012MagDownNu2.5Pythia8_Sim08e_Digi13_Trig0x409f0045_Reco14a_Stripping20NoPrescalingFlagged_ALLSTREAMS.DST.py')
j.backend = Dirac()
j.submit()
```
worked just fine :slightly_smiling_face:. Would be useful that a Warning/Error raises if you set an invalid project/version in `prepareGaudiExec` though.

**Comment by cmarinbe** [comment]
However the inputdata doesn't look to be picked up properly:
```
Ganga In [17]: j.inputdata = j.application.readInputData('data/MC_2012_27163003_Beam4000GeV2012MagDownNu2.5Pythia8_Sim08e_Digi13_Trig0x409f0045_Reco14a_Stripping20NoPrescalingFlagged_ALLSTREAMS.DST.py')
INFO     Started parsing input Data file
INFO     Finished parsing input Data file

[16:37:04]
Ganga In [18]: j.inputdata

```

**Comment by alexpearce** [comment]
Any ideas @mesmith75?

**Comment by mesmith75** [comment]
Try just
```python
j.application.readInputData('data/MC_2012_27163003_Beam4000GeV2012MagDownNu2.5Pythia8_Sim08e_Digi13_Trig0x409f0045_Reco14a_Stripping20NoPrescalingFlagged_ALLSTREAMS.DST.py')
```
That will set the inputdata attribute for you so you don't need the =

**Comment by cmarinbe** [comment]
Thanks, that works nicely!

Btw, has the default behaviour changed for the outputfiles? If I recall properly, with a DV application the .root files were automatically added as LocalFiles. But it looks like GaudiExec doesn't have this feature, you need to specify any output files manually. Is that right or am I missing something? Both me and @renaudin are seeing this behaviour.

**Comment by rmatev** [comment]
FWIW, the recipe to use a released project (also in `help(GaudiExec)`)
```pyhon
j=Job(application=prepareGaudiExec('DaVinci','v41r3'))
```
is a bit of a time bomb. This is because if the directory in `~/cmtuser` already exists, it will just do nothing, not even a warning.

My recommendation, for the starterkit at least, is to use the current directory (assuming this is not e.g. `~`):
```python
prepareGaudiExec('DaVinci', 'v41r3', myPath='.')
```

**Comment by rmatev** [comment]
@cmarinbe I'd say explicit is better than implicit in this case. Automagically copying files is not nice.

**Comment by cmarinbe** [comment]
Thanks for the tip, that's pretty bad. Using the current directory sounds good. We can create a dedicated one at the beginning of the lesson.
For the files, that's fair enough. Just wanted to check since our previous lesson didn't specify them and worked. But indeed, better to teach good practices :slightly_smiling_face: Thanks!

**Comment by mesmith75** [comment]
@rmatev Thanks for pointing that out - I'll add in a warning for the next release.
@cmarinbe Re. the output - this shouldn't have anything to do with GaudiExec(). I'll check it out, at some point it may have changed to file size. Also check your .gangarc in case you set the default for LocalFile / DiracFile.

**Comment by alexpearce** [comment]
Closed by #211.
